src/Microsoft.Azure.EventHubs/EventHubsDiagnosticSource.cs

```diff
@@ -0,0 +1,125 @@
+            if (!DiagnosticListener.IsEnabled(SendActivityName,
+                new
+                {
+                    Endpoint = csb.Endpoint,
+                    EntityPath = csb.EntityPath,
+                    PartitionKey = partitionKey
+                }))
+            {
+                return null;
+            }
+
+            Activity activity = new Activity(SendActivityName);
+
+            // extract activity tags from input
+            activity.AddTag("component", "Microsoft.Azure.EventHubs");
+            activity.AddTag("span.kind", "producer");
+            activity.AddTag("operation.name", $"Send");
+            activity.AddTag("operation.data", $"{csb.EntityPath}/{partitionKey}");
+            activity.AddTag("peer.service", "Azure Event Hub");
+            activity.AddTag("peer.hostname", csb.Endpoint.OriginalString);
+            activity.AddTag("eh.event_hub_name", csb.EntityPath);
+            activity.AddTag("eh.partition_key", partitionKey);
+            activity.AddTag("eh.event_count", count.ToString());
+            activity.AddTag("eh.client_id", clientId);
+
+            // in many cases activity start event is not interesting, 
+            // in that case start activity without firing event
+            if (DiagnosticListener.IsEnabled(SendActivityStartName))
+            {
+                DiagnosticListener.StartActivity(activity,
+                    new
+                    {
+                        Endpoint = csb.Endpoint,
+                        EntityPath = csb.EntityPath,
+                        PartitionKey = partitionKey,
+                        EventDatas = eventDatas
+                    });
+            }
+            else
+            {
+                activity.Start();
+            }
+
+            return activity;
+        }
+
+        internal static void FailSendActivity(Activity activity, EventHubsConnectionStringBuilder csb, string partitionKey, IEnumerable<EventData> eventDatas, Exception ex)
+        {
+            // TODO consider enriching activity with data from exception
+
+            if (!IsEnabled || !DiagnosticListener.IsEnabled(SendActivityExceptionName))
+            {
+                return;
+            }
+
+            DiagnosticListener.Write(SendActivityExceptionName,
+                new
+                {
+                    Endpoint = csb.Endpoint,
+                    EntityPath = csb.EntityPath,
+                    PartitionKey = partitionKey,
+                    EventDatas = eventDatas,
+                    Exception = ex
+                });
+        }
+
+        internal static void StopSendActivity(Activity activity, EventHubsConnectionStringBuilder csb, string partitionKey, IEnumerable<EventData> eventDatas, Task sendTask)
+        {
+            if (activity == null)
+            {
+                return;
+            }
+
+            // stop activity
+            activity.AddTag("error", (sendTask?.Status == TaskStatus.RanToCompletion).ToString());
```

Comment:
Is boolean's `ToString` need to define the culture info? You may want to use constants so there is no uncertainty whether it's `true` or `True`

Reply:
Ideally, you would only add the tag if there was an error. This way, it won't even be sent to the tracing system in case of success.

Reply:
makes sense - will set it only on failure and use constant ```"true"``` string

Reply:
error tag duplicates TaskStatus. Why do we need it?